### PR TITLE
SetRecursiveAcl refinement.

### PR DESCRIPTION
**Pull request by katmsft**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-files-datalake/CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,8 @@
 - Added `Metadata`, `AccessType`, `HasImmutabilityPolicy`, `HasLegalHold`, `LeaseDuration`, `LeaseState` and `LeaseStatus` to `FileSystemItem`.
 - Added new type `LeaseDurationType` to indicate if a lease duration is fixed or infinite.
 - Added `RequestId` in each return type for REST API calls, except for concurrent APIs.
+- Added `UpdateAccessControlRecursiveSinglePage` to update the access control recursively for a datalake directory.
+- Added `RemoveAccessControlRecursiveSinglePage` to remove the access control recursively for a datalake directory.
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/CHANGELOG.md:19]: It used to be `SetAccessControl...` right? Why rename? to align with some other SDK or suggested by somebody?

File: sdk/storage/azure-storage-files-datalake/CHANGELOG.md [hunk 2]
```diff
@@ -31,6 +33,8 @@
 - Removed unused type `AccountResourceType` and `PathLeaseAction`.
 - Changed all previous `LeaseDuration` members to a new type named `LeaseDurationType`.
 - `startsOn` parameter for `GetUserDelegationKey` was changed to optional.
+- Changed `SetAccessControlRecursive` to `SetAccessControlRecursiveSinglePage`, to mark that it is a single page operation, and removed the `mode` parameter, separated the modify/delete functionality to two new APIs.
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/CHANGELOG.md:36]: You're also going to add `SinglePage` suffix to other page operations, right?

Do we have an issue tracking that already?

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp [hunk 3]
```diff
@@ -160,24 +160,65 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+     * entry (ACE) consists of a scope, a type, a user or group identifier, and permissions.
+     * @param options Optional parameters to set an access control recursively to the resource the
+     * directory points to.
+     * @return
+     * Azure::Core::Response<Models::RemoveDataLakeDirectoryAccessControlRecursiveSinglePageResult>
+     * @remark This request is sent to dfs endpoint.
+     */
+    Azure::Core::Response<Models::RemoveDataLakeDirectoryAccessControlRecursiveSinglePageResult>
+    RemoveAccessControlRecursiveSinglePage(
+        std::vector<Models::Acl> acls,
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:215]: if `Remove`, why do we still need `acls`?

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp [hunk 4]
```diff
@@ -200,6 +241,14 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
     {
     }
+
+    Azure::Core::Response<Models::SetDataLakeDirectoryAccessControlRecursiveSinglePageResult>
+    SetAccessControlRecursiveSinglePageInternal(
+        Models::PathSetAccessControlRecursiveMode mode,
+        std::vector<Models::Acl> acls,
+        const SetDataLakeDirectoryAccessControlRecursiveSinglePageOptions& options
+        = SetDataLakeDirectoryAccessControlRecursiveSinglePageOptions()) const;
+
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:251]: I suggest you move this function to .cpp file to make this header file clearer

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp [hunk 5]
```diff
@@ -160,24 +160,65 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+     * @param acls Updates POSIX access control rights on files and directories. Each access control
+     * entry (ACE) consists of a scope, a type, a user or group identifier, and permissions.
+     * @param options Optional parameters to set an access control recursively to the resource the
+     * directory points to.
+     * @return
+     * Azure::Core::Response<Models::UpdateDataLakeDirectoryAccessControlRecursiveSinglePageResult>
+     * @remark This request is sent to dfs endpoint.
+     */
+    Azure::Core::Response<Models::UpdateDataLakeDirectoryAccessControlRecursiveSinglePageResult>
+    UpdateAccessControlRecursiveSinglePage(
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:193]: what's the difference between set and update?

File: sdk/storage/azure-storage-files-datalake/CHANGELOG.md (hunk 2 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/CHANGELOG.md:36]: Yes: https://github.com/Azure/azure-sdk-for-cpp/issues/1471

File: sdk/storage/azure-storage-files-datalake/CHANGELOG.md (hunk 1 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/CHANGELOG.md:19]: Yes, it was Ted's comment and it aligns with .NET.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 3 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:215]: Remove the specified ACLs.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 5 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:193]: I don't know, it is server behavior.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 3 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:215]: So you can remove some ACLs and keep others unchanged? You sure REST API supports this?

Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:215]: I don't know but it's what .NET does.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 4 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:251]: Then I need to add 3 API implementations in cpp file. I think it is only a personal preference so I'll just leave it here.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 3 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:215]: I was under the impression that we need to write unit test cases to cover these APIs.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 4 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:251]: okay, whatever.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 5 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:193]: so are you gonna write unit test?

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:193]: I don't understand, as developer of this SDK, we don't even know the behavior of the API we wrote?

File: sdk/storage/azure-storage-files-datalake/CHANGELOG.md (hunk 1 above)
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/CHANGELOG.md:19]: What are we gonna do about path client?

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]
I think it's basic requirements for developers of storage sdk to understand the semantics of storage API. But evidently it's not the case yet for the three APIs we changed in this PR. So I'd recommend we do more investigation first and then continue the discussion based on clearer understanding of the semantics and behavior of these APIs and the differences among them.

File: sdk/storage/azure-storage-files-datalake/CHANGELOG.md (hunk 1 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/CHANGELOG.md:19]: Nothing, path client only contains API work for both directory/file, while this API is not the case.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 5 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:193]: Sure, I was trying to do it fast due to the deadline we are catching, but I agree that not understanding the API clearly is not acceptable, I'll reset my attitude.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 3 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:215]: Added.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp (hunk 5 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_directory_client.hpp:193]: `Update` will only be setting the matching type/ID of the ACL, while `Set` will overwrite all ACLs.
`Delete`, as I previously guessed but not confirmed, is confirmed to be only deleting the specified ACL.

**Comment by Jinming-Hu** [comment]
In .Net, `PathClient` has `{Set,Update,Remove}AccessControlRecursive` and `SetAccessControlList`. `DirectoryClient` inherits these functions and override `SetAccessControlList`

what's the consideration behind this unalignment?

**Comment by katmsft** [comment]
> In .Net, `PathClient` has `{Set,Update,Remove}AccessControlRecursive` and `SetAccessControlList`. `DirectoryClient` inherits these functions and override `SetAccessControlList`
> 
> what's the consideration behind this unalignment?

What does it mean if a file resource is initialized as a path client and `AccessControlRecursive` related API is called? It does not make sense.

We can always move it into the path client in the future as a non-breaking change.

**Comment by Jinming-Hu** [comment]
> What does it mean if a file resource is initialized as a path client and `AccessControlRecursive` related API is called? It does not make sense.

What does it mean if a file resource is initialized as a path client and `Delete` with `recursive = true` is called?

**Comment by katmsft** [comment]
> > What does it mean if a file resource is initialized as a path client and `AccessControlRecursive` related API is called? It does not make sense.
> 
> What does it mean if a file resource is initialized as a path client and `Delete` with `recursive = true` is called?

That's different, a meaningless combination of the parameters and a meaningless API itself are different levels of exposure to the customer.
If you insist, I can delete the `delete` methods in the path client.

**Comment by katmsft** [comment]
Force push to resolve the merge issue caused by github.

**Pull request closed without merging**